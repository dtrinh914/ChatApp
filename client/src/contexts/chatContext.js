import React, {createContext, useReducer} from 'react';
import chatReducer from '../reducers/chatReducer';

const initialState = {user:"", groups:"", selected:""}

export const ChatContext = createContext();

export function ChatProvider(props){
    const [chatData,chatDispatch] = useReducer(chatReducer, initialState);
    return(
        <ChatContext.Provider value={{chatData, chatDispatch}}>
            {props.children}
        </ChatContext.Provider>
    )
}