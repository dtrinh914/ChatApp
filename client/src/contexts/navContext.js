import React, {createContext, useReducer} from 'react';
import navReducer from '../reducers/navReducer';

const initialState = {
    rightNav:{
        root:false,
        addMem:false
    }
}

export const NavContext = createContext();

export function NavProvider(props){
    const [navData , navDispatch] = useReducer(navReducer, initialState);
    return(
        <NavContext.Provider value={{navData, navDispatch}}>
            {props.children}
        </NavContext.Provider>
    )
}