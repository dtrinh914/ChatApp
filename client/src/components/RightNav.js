import React, {useContext} from 'react';
import Paper from '@material-ui/core/Paper';
import Hidden from '@material-ui/core/Hidden';
import {makeStyles} from '@material-ui/styles';
import AddMember from './AddMember';
import {NavContext} from '../contexts/navContext';
import {ChatContext} from '../contexts/chatContext';
import axios from 'axios';

const useStyles = makeStyles({
    root:{
        display: 'flex',
        flexDirection: 'column',
        width: '250px',
        height: '100%',
        border: 'none',
        background:'#424242',
        color: 'white',
        overflow: 'auto'
    },
    divider:{
        width: '85%',
        backgroundColor: '#616161'
    }
});


export default function RightNav({updateInvite, updateMembers}) {
    const {navData} = useContext(NavContext);
    const {chatData, chatDispatch} = useContext(ChatContext);
    const classes = useStyles();
    const addMemStatus = navData.rightNav.addMem;

    const selectedGroupId = chatData.selected._id;
    const currUserId = chatData.user._id;
    const {activeMembers, pendingMembers, pendingRequests} = chatData.groups[chatData.selected.index];

    const sendInvite = (userId) =>{
        axios.post('/api/groups/' + selectedGroupId + '/members',{userId: userId, 
            withCredentials:true})
            .then(res => {
                if(res.data.status === 1){
                    chatDispatch({type:'ADD_MEMBER', groupId:selectedGroupId, memberId: userId});
                    updateInvite(userId);
                    updateMembers(selectedGroupId);
                }
            })
            .catch(err => {
                console.log(err);
            })
    }

    //filter the search results based on if the user is the current user, already a member,
    //or has a pending invite
    const filterResults = (results) => {
        let filtered = [];
        
        for(let i = 0; i < results.length; i++){
            let current = results[i];
            if(current._id === currUserId){
                continue;
            }
            //check if current user is an active member 
            //and creates a key 'status' set to 'active' for the user object
            if(activeMembers.includes(current._id)){
                current['status'] = 'active';
            }
            //check if current user had a pending request for the group
            else if(pendingMembers.includes(current._id) || pendingRequests.includes(current._id)){
                current['status'] = 'pending';
            } else {
                current['status'] = 'add';
            }
            filtered.push(current)
        }
        return filtered;
    }


    return (
        <Hidden smDown >
            <Paper className={classes.root} square>
                <AddMember open={addMemStatus} sendInvite={sendInvite} filterResults={filterResults} />
            </Paper>
        </Hidden>
    )
}
