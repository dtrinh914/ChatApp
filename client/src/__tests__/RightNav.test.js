import React from 'react';
import RightNav from '../components/RightNav';
import {ChatProvider} from '../contexts/chatContext';
import {NavProvider} from '../contexts/navContext';
import {render, cleanup} from '@testing-library/react';
import {createMuiTheme} from '@material-ui/core/styles';
import {MuiThemeProvider} from '@material-ui/core/styles';

afterEach(cleanup);

const seedData = {"_id":"5e2a008c295fb74748745ab9",
                  "groupName":"New group",
                  "description":"",
                  "activeMembers":[
                                  {"_id":"5e28c5c185ffc434b337c266","username":"test_user1"},
                                  {"_id":"5e28c5bc85ffc434b337c264","username":"test_user2"},
                                  {"_id":"5e28c5c485ffc434b337c268","username":"test_user3"}
                                  ],
                   "pendingMembers":[],
                   "pendingRequests":[],
                   "blocked":[],
                   "creator":"5e28c5c185ffc434b337c266",
                   "admins":["5e28c5c185ffc434b337c266"],
                   "memberMap":{
                               "5e28c5c185ffc434b337c266":{"username":"test_user1"},
                               "5e28c5bc85ffc434b337c264":{"username":"test_user2"},
                               "5e28c5c485ffc434b337c268":{"username":"test_user3"}
                               }
                 }

const component = <ChatProvider>
                  <NavProvider>
                        <RightNav currentGroup={seedData} />
                  </NavProvider>
                  </ChatProvider>

describe('right nav when window size is md', ()=>{
    const theme = createMuiTheme({props:{MuiWithWidth: {initialWidth:'md'}}}); 
                        
    it('should render properly and match snapshot', ()=>{
        const {container, queryByTestId} = render(<MuiThemeProvider theme={theme}>
                                                        {component}
                                                  </MuiThemeProvider>);
        expect(queryByTestId('rightnav-drawer')).toBeNull();
        expect(container).toMatchSnapshot();
    });
})

describe('right nav when window size is sm', ()=>{
    const theme = createMuiTheme({props:{MuiWithWidth: {initialWidth:'sm'}}});
    it('should render empty div and match snapshot', ()=>{
        const {container} = render(<MuiThemeProvider theme={theme}>
                                        {component}
                                   </MuiThemeProvider>);
        expect(container).toMatchSnapshot();
    });
})
