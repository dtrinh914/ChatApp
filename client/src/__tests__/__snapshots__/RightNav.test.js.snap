// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`right nav when window size is md should render properly and match snapshot 1`] = `
<div>
  <div
    class="makeStyles-root-1"
    data-testid="right-nav"
  >
    <div
      class="makeStyles-paper-3"
    >
      <div>
        <p
          data-testid="group-description"
        >
          [No description has been written]
        </p>
      </div>
      <hr
        class="MuiDivider-root makeStyles-divider-4 MuiDivider-middle"
      />
      <div>
        <h2>
          Group Members
        </h2>
        <ul
          class="MuiList-root makeStyles-list-12 MuiList-padding"
        >
          <li
            class="MuiListItem-root MuiListItem-gutters"
            data-testid="group-member"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-13 MuiAvatar-colorDefault"
            >
              T
            </div>
            <span
              class="makeStyles-badgeOutline-36"
            />
            <span
              class="makeStyles-badgeCore-37"
              style="background: rgb(0, 230, 118);"
            />
            <span>
              test_user1
            </span>
          </li>
          <li
            class="MuiListItem-root MuiListItem-gutters"
            data-testid="group-member"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-13 MuiAvatar-colorDefault"
            >
              T
            </div>
            <span
              class="makeStyles-badgeOutline-36"
            />
            <span
              class="makeStyles-badgeCore-37"
              style="background: rgb(0, 230, 118);"
            />
            <span>
              test_user2
            </span>
          </li>
          <li
            class="MuiListItem-root MuiListItem-gutters"
            data-testid="group-member"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-13 MuiAvatar-colorDefault"
            >
              T
            </div>
            <span
              class="makeStyles-badgeOutline-36"
            />
            <span
              class="makeStyles-badgeCore-37"
              style="background: rgb(0, 230, 118);"
            />
            <span>
              test_user3
            </span>
          </li>
        </ul>
      </div>
    </div>
  </div>
  
  
  
</div>
`;

exports[`right nav when window size is sm should render empty div and match snapshot 1`] = `
<div>
  
  
  
</div>
`;
